### ISSUE TITLE
Proposed new workspace layout on disk

### ISSUE THREAD

**Issue by evansd** [opensafely/job-runner#72]
Following on from @ghickman's work on creating a workspace-based UI for the job-server I wonder if it would make sense to have something similar on the job-runner.

Example:
```
my-workspace-name/

  outputs/
    analysis/
      logfile.log
    output/
      input.csv
      input.dta
      chart.png

  output-manifest.json

  generate_cohort.log
  prepare_data.log
  make_chart.log
```

The `ouputs` directory contains all the output files (of the appropriate privacy level) from all the actions run so far, laid out as they would be if each action had been run in the same working directory outside the job-runner.

`output-manifest.json` looks something like this:
```
{
  "outputs": {
    "analysis/logfile.log": {"created_by_action": "make_chart"},
    "output/input.csv": {"created_by_action": "generate_cohort"},
    "output/input.dta": {"created_by_action": "prepare_data"},
    "output/chart.png": {"created_by_action": "make_chart"},
  },
  "actions": {
    "generate_cohort": {
       "commit": "807c77526f85a91bb41ee6e24533360e5d047931"
       "job_id": "xyz123",
      ...
    },
    ...
  }
}
```
That is, for each file it lists the action that generated it, and then for each action it gives various details about how it was run, who ran it etc.

The log files are named `<action_name>.log` and contain the stdout/stderr from each job. It might also make sense for us to "top and tail" these logs with some job-runner specific log lines.

My hypothesis is that while this is in one sense functional equivalent to what we have now (it's the same set of files, just organised differently) it will be much easier for researchers to understand and work with.

**Comment by ghickman** [comment]
I'm not sure how we lay out files currently but the directory structure immediately made me think of the `project.yaml`s I've seen and I think that's a great Pro in it's favour!

Small nit: I think `outputs/output` is going to trip us up while debugging things and researchers while they onboard.

**Comment by evansd** [comment]
Yes, `outputs/output` is painful. I was thinking of calling the directory `files`. However the fact that we have an `output` directory at all is really just an artefact of how the cohortextractor happens to work at the moment. Whereas the key in project.yaml under which output files are defined is called `outputs`. So I'd rather keep the the first `outputs` and do something about the second `output`, if that makes any kind of sense? Maybe change it to `outputs/data` or something.

**Comment by ghickman** [comment]
That all makes sense 👍

**Comment by alexwalkercebm** [comment]
I agree that `output` is awkward and is the one that should change. There's a fair bit of Stata code that would need to be changed if it did. Both reading `output/input*.csv` and saving intermediate `.dta` files and other stuff to `output/`.

Is there a specific name that will be used for the folder that L4 outputs are saved to? Or are they manually added to the repo by the reviewer? When doing this on the L2/3, we've been making a new folder called something like `released_analysis_results`.

**Comment by evansd** [comment]
> There's a fair bit of Stata code that would need to be changed if it did.

If we did change the directory it would be a new default for future projects, but we'd have to do it in a way that didn't break the existing ones. Seb had the foresight to add a version number to the project.yaml format so it shouldn't be too difficult to make this kind of change while still allowing old projects to run.

> Is there a specific name that will be used for the folder that L4 outputs are saved to? Or are they manually added to the repo by the reviewer?

I don't think there's any specific special handling of L4 outputs at the moment. We could make it a convention, and we could enforce that convention in the job-runner, that they should always go in a specific directory. But I haven't given that a lot of thought yet (Seb may have done though).

**Comment by sebbacon** [comment]
This looks good. The one thing that's not clear to me is how this works across the two privacy levels where a single action writes to both. I do like the simplicity (from a security POV) of the current setup where we know they are always on completely different volumes, though this is not strictly necessary.   Are you perhaps suggesting they would be merged into one directory structure on the L3 server, and we'd use a different process from the current one to copy some of these to L3?  

To illustrate / think aloud, consider:

```yaml
  foo:
    run: python:latest foo.py
    outputs:
      moderately_sensitive:
        thing: thing.txt
     highly_sensitive:
        thing2: thing2.txt
```
With the current way of dividing L3 and L4 we'd end up with this:

L3:
```
highly_sensitive/
  my-workspace-name/
    outputs/
      thing2.txt
    output-manifest.json
    foo.log
```
L4:
```
moderately_sensitive/
  my-workspace-name/
    outputs/
      thing.txt
    output-manifest.json
```

If we combined on L3 we would have this on L3:
```
highly_sensitive/
  my-workspace-name/
    outputs/
      thing.txt
      thing2.txt
    output-manifest.json
    foo.log
```

And on L4:
```
moderately_sensitive/
  my-workspace-name/
    outputs/
      thing.txt
    output-manifest.json
```

**Comment by evansd** [comment]
I think having them combined is the way to go, so the high privacy directory contains everything, the medium contains medium and low, and the low just contains low.

We can still use separate volumes for the different privacy levels.

This is obviously a bit less efficient because we have to make multiple copies of things, but my thought was that in the majority of cases the really big files are going to be the high privacy ones so we they'll only exist in one place anyway.

**Comment by evansd** [comment]
Further burst of inspiration (or what seems to me to be inspiration :)

Instead of having the workspace contain a bunch of metadata and log files and then an `outputs` directory with output files we could invert this and put all the metadata and logs in a directory along with the outputs. If we called it, for want of a better name right now, `opensafely` then the structure would look something like this:
```
my-workspace-name/

  analysis/
    logfile.log
  output/
    input.csv
    input.dta
    chart.png

  opensafely/
    output-manifest.json
    generate_cohort.log
    prepare_data.log
    make_chart.log
```

Of course, it's a bit inelegant to mush together the two sets of outputs like this _but_ this has the big advantage that we can make it work in exactly the same way locally i.e. when you use the local run command you get your output files wherever your code puts them and an `opensafely` directory with logs and other metadata.

This means that researchers can get used to running their code locally and then know exactly what to expect, and where, when they run it through the job server. In fact, it lets us tell the story that a "workspace" is sort of like a checkout of your code on the server which you interact with via the job-server. (I've got some further ideas on how we could push this illusion even further but I don't want to muddy the discussion with them now.)

I think this could really help in reducing the (inevitably steep) learning curve.

**Comment by sebbacon** [comment]
This is a great idea. My only question is how we might indicate the difference between high / medium / low privacy outputs on disk.  No bright ideas right now though!

**Comment by ghickman** [comment]
This all makes a lot of sense to me!  I particularly liked `This means that researchers can get used to running their code locally` 👍

As for the colour of the bikeshed, could `metadata` work better than `opensafely`?  On the one hand it's less specific but my worry would be that `opensafely` is too specific.  Thoughts?

**Comment by CarolineMorton** [comment]
Agreed with @ghickman `opensafely` may be too specific. `metadata` is ok

**Comment by evansd** [comment]
Implemented in #71
